Spatial raster prep (#7)
ninoxconsulting/r-telemetry-workshop-nov-2023: 08-rsf-prep.qmd

---
title: "Generate random sample points for RSF"
---

## Overview:

In this module we use our prepared spatial data to generate the input datafiles to conduct a resource selection function analysis. This includes: 

-   generate a set of "available" or "background points" from our study area.

> -   extract spatial information for presence and available point locations

Reviewer: Curious (not really relevant to the course) - theoretically, if all the animals in a herd are tagged, can you treat those available point locations as true absences (since you know the animal wasn't there?)

While in this example we are concentrating on future analysis, the process of extracting background information for known locations can also provide meaningful summary statistics, i.e. proportion of locations within a specific BEC zone.

## 1. Generate availalble location points.

We will generate a simple set of "available" based on the geographic distribution of the study area. Note you can limit this area in which you generate "background" points, using more sophisticated methods, such as within a kernal density or home range estimate.

Lets start by reading in the libraries we will use.

```{r, eval = FALSE}

library(dplyr)
library(terra)
library(sf)
library(mapview)
library(ggplot2)

```

Next, read in the standardized data points for the Scott herd. If needed we can transform to BC Albers projection (EPGS:3005) to match our raster stack prepared in the previous step. 

```{r, eval = FALSE}

# read in the aoi template 

template <- rast(file.path("clean_data","template.tif"))

# read in points 

> bou_pts <- st_read(file.path("clean_data/scott_herd_subset.gpkg")) 

Reviewer: For me this is only a csv currently

Author: ok roger, i think i need to rerun from the start to clear out the old stuff.

bou_pts <- st_transform(bou_pts, 3005)

# Lets keep only the important columns and add a "presence/absence" column. 

bou_pts <- bou_pts %>% 
  dplyr::select(animal.id, jdate)%>%
  mutate(pres_abs = 1)

```

We can use the **spatSample** function to generate random points for our given study area. This function has many more options which can be reviewed by using : **?spatSample** in the console. 

Lets generate a set of points the same length as our "presence" locations using a "random" method. 

```{r, eval = FALSE}

# Generate random points for RSF use areas.
set.seed(123)
avail_points <- spatSample(template, size = 2906, as.points = TRUE, na.rm = TRUE, method = "random")

avail_points <- st_as_sf(avail_points)

# lets rename the column to make it clear these are background points 
avail_points <- avail_points %>%

>   rename("pres_abs" = lyr.1 )

Reviewer: When I do this locally the column is called `aoi`... not sure if that's a package version thing, or a data-flow thing... we can come back to it when we nail down the data flow

```

> We can do a quick review fo the points to see what they look like using mapview.  

Reviewer: ```suggestion
We can do a quick review of the points to see what they look like using mapview.  
```

```{r, eval = FALSE}

mapview(avail_points) +
mapview(bou_pts, color = "red", cex= 3)

```

We can now combine our caribou locations and "available" locations into a single dataset. We will retain the spatial information to allow us to easily extract the values in the corresponding raster stack 

```{r, eval = FALSE}

avail_points 

bou_pts

# note we have slightly different column headers "geom" vs "geometry" 

> st_geometry(avail_points) = "geom" 

Reviewer: I didn't need to do this, for me they were both called `geometry`. Again, probably a data version thing

allpts <- bind_rows(bou_pts, avail_points ) 

```

> Next, we can read in our prepared raster stack as an .rds object. We can now use the **extract** function from the **terra** package to extract information for all layers in the raster stack for each of our points.  

Reviewer: ```suggestion
Next, we can read in our prepared raster stack from the .Rds file we saved previously. We can now use the **extract** function from the **terra** package to extract information for all layers in the raster stack for each of our points.  
```

```{r, eval = FALSE}

# read in the raster stack 

rstack <- readRDS(file.path("clean_data", "covars.RDS"))

# extract all values in the raster stack for each location in the bou_pts file. 

atts <- terra::extract(rstack, allpts)

# remove unused columns 

bou_full_pts <- cbind( allpts, atts) %>%
  select(-ID)

> # export this as spatial file # or 

Reviewer: Could show how to use `st_write()` and explain the `cbind(st_coordinates())` bit

Author: add note to check this

bou_table <- bou_full_pts %>%
  cbind(st_coordinates(.)) %>%
  st_drop_geometry()

# write out as csv 

write.csv(bou_table, file.path("clean_data", "allpts_att.csv"))

```

# Summary of the extracted Point Values

```{r, eval = FALSE}

# tri - good overlap

ggplot(bou_table, aes(aspect)) +
  geom_histogram(binwidth = 20) +
  facet_wrap(~pres_abs)

> ggplot(bou_table, aes(lyr.1)) +

Reviewer: The column is called `aoi` for me... see comment above

Author: yes I agree this is a legacy issue
